tokio/src/runtime/thread_pool/worker.rs

```diff
@@ -466,11 +478,17 @@ impl Context {
         *self.core.borrow_mut() = Some(core);
 
         // Park thread
+        if let Some(f) = &self.worker.shared.before_park {
+            f()
+        }
         if let Some(timeout) = duration {
             park.park_timeout(timeout).expect("park failed");
```

Comment:
hmm. if there's a timeout, i wonder if we want to include the duration of the `before_park` callback in the timeout? in some cases, running the callback may be time consuming (e.g. when QSBR or a similar technique is being used to run a lot of low-priority work when parking).

we might want to consider taking a timestamp prior to running the callback, and before parking, we check if the entire duration has elapsed running the callback. if so, return without parking. otherwise, we could park for the provided duration _minus_ the time spent running the callback.

cc @carllerche do you think it makes sense to include time spent in the callback as part of the park duration?

Reply:
In particular, the timeout might be zero in case we are running maintenance and not actually parking.

Reply:
The timer driver generally wraps the parker, and will update the provided timeout from `None` to the smallest timer as appropriate.

Reply:
I rearranged for other reasons in db54bb7, so I no longer depend on timeout being None for parked threads waiting on a timer.